### PR TITLE
Allow configuration of fields during register

### PR DESCRIPTION
**Pull request by derrickmehaffy**
What it does

Modifies users-permissions to pick fields from registration body instead of omitting them and allows for plugin configuration to let users select what fields are allowed during the registration phase.

How to test

Configure the plugin like the following:

```js
// path: ./config/plugins.js

module.exports = {
  'users-permissions': {
    config: {
      register: {
        allowedFields: [ 'username', 'password', 'email', 'role' ]
      }
    }
  }
}
```

This should allow you to specify the role ID during the registration phase (not that I'd recommend that personally)

[Documentation PR](https://github.com/strapi/documentation/pull/1703)

### CODE REVIEW

File: packages/plugins/users-permissions/server/controllers/auth.js
Comment (christiancp100) [inline packages/plugins/users-permissions/server/controllers/auth.js:281]: What's this `register?.allowedFields`?

Comment (derrickmehaffy) [reply packages/plugins/users-permissions/server/controllers/auth.js:281]: New plugin configuration option, basically checking if the user has modified it and if not, using the defaults.

Instead of forcing the user to only accept what we allow, it grants them the freedom to configure it themselves.

Comment (innerdvations) [reply packages/plugins/users-permissions/server/controllers/auth.js:281]: Should 'picture' be allowed by default?

Comment (AnnikenYT) [reply packages/plugins/users-permissions/server/controllers/auth.js:281]: > Should 'picture' be allowed by default?

Picture isn't a default field on user, is it?

Comment (innerdvations) [reply packages/plugins/users-permissions/server/controllers/auth.js:281]: You're right, I was looking at getstarted 🤦

Comment (simotae14) [inline packages/plugins/users-permissions/server/controllers/auth.js:281]: you can avoid to check again register?.allowedFields because it is compiled in another check

File: packages/plugins/users-permissions/server/controllers/auth.js
Comment (innerdvations) [inline packages/plugins/users-permissions/server/controllers/auth.js:281]: I wouldn't allow overriding the three required fields, we should make it "additionalFields" and merge it with username, email, password.

**Review by innerdvations — CHANGES_REQUESTED** [review]
This could be considered a breaking change, since users who have modified their user schema may be expecting those additional fields to be set on register.

I think it would be acceptable in this case though if we hold this until the next minor release and add a note in the release log about it so anyone doing so can add this setting.

File: packages/plugins/users-permissions/server/controllers/auth.js
Comment (christiancp100) [inline packages/plugins/users-permissions/server/controllers/auth.js:281]:
```suggestion
        register?.allowedFields ?? ['username', 'email', 'password']
```

**Review by christiancp100 — CHANGES_REQUESTED** [review]

File: packages/plugins/users-permissions/server/controllers/auth.js
Comment (Convly) [reply packages/plugins/users-permissions/server/controllers/auth.js:281]: Agree with Ben here

**Review by Convly — CHANGES_REQUESTED** [review]

**Review by innerdvations — APPROVED** [review]
Before merging, we need to have documentation written and then we need to put a "breaking change" in the patch notes instructing anyone with custom user fields they expect to be set on registration to be added to the config.

But I think the breaking change is worth it in this case, it only breaks code for users who are relying on a vulnerability in their code. We need to wait for at least a minor version though, not the next patch version.

**Comment by innerdvations** [comment]
We need to be sure to mark this as a breaking change in 4.11 and sync publication of the documentation.

File: packages/plugins/users-permissions/server/controllers/auth.js
Comment (Convly) [inline packages/plugins/users-permissions/server/controllers/auth.js:12]:
```suggestion
const { concat, compact, uniq } = require('lodash/fp');
```

File: packages/plugins/users-permissions/server/controllers/auth.js
Comment (Convly) [inline packages/plugins/users-permissions/server/controllers/auth.js:282]:
```suggestion
        uniq(compact(concat(['username', 'email', 'password'], register?.allowedFields)))
```

**Review by Convly — CHANGES_REQUESTED** [review]

**Review by Convly — APPROVED** [review]

**Review by christiancp100 — APPROVED** [review]

**Comment by alexandrebodin** [comment]
We can make the breaking change for v5. In the meantime we can make the ommit configurable so users can still make this better without a breaking change

**Comment by NicoHof** [comment]
Hello! 
I got here because I am a bit confused by the way this breaking change is communicated.
I noticed this migration guide https://docs.strapi.io/dev-docs/migration-guides --> https://docs.strapi.io/dev-docs/migration/v4/migration-guide-4.10.1-to-4.11.0 

It says this breaking change **is already introduced in 4.11.0** but this PR is still open.

I think it is confusing for people to upgrade to 4.11.0 because custom fields still work on registration in 4.11.0 without the suggested configuration changes.

Best, Nico

**Comment by innerdvations** [comment]
> It says this breaking change **is already introduced in 4.11.0** but this PR is still open.
> 

You're right, I'll make sure the documentation gets reverted.

As far as how to proceed, here is what we've decided to do, which will require updating this PR:

- Continue to use `omit` when `allowedFields` is not present in config
  - Add pseudo-deprecation warning, on strapi startup U&P will emit a warning when `allowedFields` is not set saying that in v5 custom user fields will not be accepted for user registration unless added to `allowedFields`
- When `allowedFields` is present in config, we will switch to `pick` merging in the user's array
- New projects will be generated with `allowedFields` set to an empty array
- in v5, we will remove the `omit` option entirely (and then we don't need to require `allowedFields`, we just treat undefined as an empty array like the current PR)

**Pull request closed without merging**
